lib/solargraph/source/chain/call.rb

```diff
@@ -57,17 +57,30 @@ def inferred_pins pins, api_map, context, locals
             overloads = p.signatures
             # next p if overloads.empty?
             type = ComplexType::UNDEFINED
-            overloads.each do |ol|
+            # start with overloads that require blocks; if we are
+            # passing a block, we want to find a signature that will
+            # use it.
+            sorted_overloads = overloads.sort { |ol| ol.block? ? -1 : 1 }
+            sorted_overloads.each do |ol|
               next unless arguments_match(arguments, ol)
-              # next if ol.parameters.last && ol.parameters.last.first.start_with?('&') && ol.parameters.last.last.nil? && !with_block?
               match = true
               arguments.each_with_index do |arg, idx|
                 param = ol.parameters[idx]
+                atype = nil
                 if param.nil?
-                  match = false unless ol.parameters.any?(&:restarg?)
+                  last_arg = idx == arguments.length - 1
+                  match = if ol.parameters.any?(&:restarg?)
+                            true
+                          elsif last_arg && ol.block?
+                            # block argument that isn't declared as an arg as well
+                            atype ||= arg.infer(api_map, Pin::ProxyType.anonymous(context), locals)
+                            atype.namespace == 'Proc'
```

Comment:
Is `atype.namespace == 'Proc'` a noop?

Reply:
No - confusing logic, but it creates a boolean that flows up and gets assigned to the `match` variable